// yummyecs/dist/Framework/Pool/PublicAE.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DestroySystemAction = exports.CreateSystemAction = exports.DestroyAllActorAction = exports.DestroyActorAction = exports.CreateActorAction = void 0;
const UE = require("ue");
const AE_1 = require("../AE");
class CreateActorAction extends AE_1.Action {
    constructor(ctor, transform = new UE.Transform()) {
        super();
        this.ctor = ctor;
        this.transform = transform;
    }
}
exports.CreateActorAction = CreateActorAction;
class DestroyActorAction extends AE_1.Action {
    constructor(actor) {
        super();
        this.actor = actor;
    }
}
exports.DestroyActorAction = DestroyActorAction;
class DestroyAllActorAction extends AE_1.Action {
    constructor() {
        super();
    }
}
exports.DestroyAllActorAction = DestroyAllActorAction;
class CreateSystemAction extends AE_1.Action {
    constructor(ctor) {
        super();
        this.ctor = ctor;
    }
}
exports.CreateSystemAction = CreateSystemAction;
class DestroySystemAction extends AE_1.Action {
    constructor(ctor) {
        super();
        this.ctor = ctor;
    }
}
exports.DestroySystemAction = DestroySystemAction;
